bitbucket server api rate limiting

kjjuno opened the pull request:
## Hey, I just made a Pull Request!

Adds the ability to rate limit calls to the bitbucket server api. If no `rateLimit` is provided in the configuration no rate limiting will be applied.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

kjjuno commented:
> ## Missing Changesets
> The following package(s) are changed by this PR but do not have a changeset:
> 
> * **@backstage/integration**
> 
> See [CONTRIBUTING.md](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets) for more information about how to add changesets.
> 
> ## Changed Packages
> Package Name	Package Path	Changeset Bump	Current Version
> @backstage/integration	packages/integration	**none**	`v1.11.0`
> @backstage/plugin-catalog-backend-module-bitbucket-server	plugins/catalog-backend-module-bitbucket-server	**minor**	`v0.1.33-next.0`

I'm not sure how the `@backstage/integration` package is showing as changed. I didn't think I had changed it. Is this a mistake because of the merge conflicts I dealt with, or am I legitimately missing something here?

vinzscam reviewed (changes requested):
hey, question: are you getting the rate limiting when using the `BitbucketServerEntityProvider`? 🤔 

Looking at the code it doesn't seem you should get any rate limiting

kjjuno commented:
> hey, question: are you getting the rate limiting when using the `BitbucketServerEntityProvider`? 🤔
> 
> Looking at the code it doesn't seem you should get any rate limiting

We added this plugin to scrape our bitbucket server instance, and almost immediately we started getting performance issues on our server. The team that owns that has asked us to make sure we only hit the api around 1 time per second.

This PR adds the ability rate limit those calls by making use of the `Bottleneck` library.

The rate limiting is implemented in the `BitbucketServerClient` itself.

Does this address your concern, or do you still have questions or concerns?

vinzscam commented:
> BitbucketServerClient

my question was more about: are you using the `BitbucketServerEntityProvider` or are you using the `BitbucketServerClient` directly?

kjjuno commented:
we are using the BitbucketServerEntityProvider

Rugvip commented on plugins/catalog-backend-module-bitbucket-server/src/lib/BitbucketServerClient.ts:
```diff
@@ -28,6 +29,7 @@ import { BitbucketServerProject, BitbucketServerRepository } from './types';
  */
 export class BitbucketServerClient {
```
It's kinda nice to implement this functionality within this client directly, but right now there's a bit of a trap with that in that most implementations create this client themselves. That means that the actual number of requests hitting the Bitbucket deployment might be significantly higher than the configured throttling. In the end I think it might be best to implement this in the Bitbucket discovery provider instead? it also reduces the risk that this negatively impacts requests in the user path, like scaffolder actions.

cal5barton replied:
With this suggestion, would the throttling config options be moved to the discovery provider as well?

Rugvip replied:
Yep I think that would be the best place to put them to avoid confusion

kjjuno replied:
Ok, I've switched over to p-throttle. I have also moved the throttling logic into the entity provider as well. I believe this should be ready for another look.

Rugvip reviewed (changes requested):
Thank you! 👍

freben commented on plugins/catalog-backend-module-bitbucket-server/src/providers/BitbucketServerEntityProvider.ts:
```diff
@@ -174,13 +177,35 @@ export class BitbucketServerEntityProvider implements EntityProvider {
+      });
+
+      parseFunction = throttle((opt: ParseOptions) => {
+        return this.parse(opt);
```
I'm not sure but, is this really throttling what it should throttle? The default parse function doesn't do anything at all, and if it were only about the parse function then the caller could be responsible themselves for passing in an already throttled callback. And what if that callback performs half a million requests per call, how does that affect the throttling?

Isn't it the case that, what we want to throttle is really all `client` requests that we make? Maybe I misunderstand, but this looked surprising to me.

kjjuno replied:
It's my understanding that part of the parse function is that it makes a hit to the Bitbucket Server API to look at the configured url for matching entities. I calls to list the repositories, but this seems to be the place where the actual lookup against files in bitbucket happens.

kjjuno replied:
I understand better now. I have moved the throttling to be around the `fetch` function in the client and url readers

freben commented on plugins/catalog-backend-module-bitbucket-server/src/providers/BitbucketServerEntityProvider.ts:
```diff
@@ -204,19 +229,51 @@ export class BitbucketServerEntityProvider implements EntityProvider {
+    const now = Date.now();
+    if (this.lastCalled !== null) {
+      const duration = now - this.lastCalled;
+      this.logger.debug(`Time since last call: ${duration}ms`);
```
Hmm lots of debugging here, do we really need to keep it?

kjjuno replied:
removed

kjjuno commented on plugins/catalog-backend-module-bitbucket-server/src/providers/BitbucketServerEntityProvider.ts:
```diff
@@ -174,13 +177,35 @@ export class BitbucketServerEntityProvider implements EntityProvider {
+      | undefined;
+
+    if (this.config.throttling) {
+      const throttle = pThrottle({
```
moved to the constructor

kjjuno replied:
I've moved the throttling back to the client.

kjjuno commented on .changeset/silent-carpets-explode.md:
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/plugin-catalog-backend-module-bitbucket-server': minor
```
done

kjjuno commented on .changeset/silent-carpets-explode.md:
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/plugin-catalog-backend-module-bitbucket-server': minor
+'@backstage/integration': minor
```
done

kjjuno commented:
@freben how would you feel about going back to the approach of putting the throttling into the client? We could make it so that you need to opt into that behavior. If there is no throttling configured no throttling would be applied.

The client itself seems to be the only place I can guarantee that throttling is properly applied

kjjuno commented:
I have updated this PR based on feedback in the last framework SIG. I think this should be ready for another review

This change moves the throttling to the `fetch` function which is retrieved from the `FetchService` based on the integration config. This is used from both the Bitbucket Server Client, as well as the Url reader. The same change has been applied to the Bitbucket Cloud client and url reader.

kjjuno commented:
this shouldn't be stale. The maintainer group is on holiday at the moment.

kjjuno commented:
Still waiting for the backstage team to get back from vacation. This should not be marked stale.

brunooliveiramac commented:
thumbs up for this

kjjuno commented:
Any chance I could get a new review? This is a critical enhancement for us. If there are still any outstanding concerns please let me know.

cal5barton commented on plugins/bitbucket-cloud-common/package.json:
```diff
@@ -42,11 +42,13 @@
   "dependencies": {
     "@backstage/integration": "workspace:^",
-    "cross-fetch": "^4.0.0"
+    "cross-fetch": "^4.0.0",
```
Are these references needed here since we now reference the FetchService in the integration dependency?

kjjuno replied:
This has been removed

kjjuno replied:
Removing these broke the build. I'm putting them back.

benjdlambert reviewed (changes requested):
Hey :wave:

Apologies for the long turnaround on reviews for this, we've been pretty slammed after the holiday season, so we're just getting back to full capacity again!

I'm thinking that whilst we might get to having some form of `FetchService` in the backend, similar to the `fetchApiRef` in the frontend, we're not really there yet and we don't know what that's going to look like either. I don't want to be even more of a blocker for this work, so what I would suggest is that we narrow down the changes for what's required for this PR just to the `@backstage/plugin-bitbucket*` packages instead, and leave out the other changes for now.

It might need some restructuring of course, to maybe have this in some other config variable or to be set in some other way in the interim and out of the `integrations.config` for now, but I think that it's probably easier to go down that path, and then eventually deprecate that way once we have some more thoughts about what we're going to do with a global `FetchService`.

Does this make sense for now?

joshjung commented on packages/integration/api-report.md:
```diff
@@ -187,6 +191,7 @@ export type BitbucketCloudIntegrationConfig = {
   username?: string;
   appPassword?: string;
   token?: string;
+  throttling?: ThrottlingConfig;
```
I really like this approach, but in line with @benjdlambert suggestion of making this more consistent across services, I am curious if perhaps we put the `throttling` key underneath of a `fetch` key, like so:

```
fetch:
  throttling:
    count: 1
    interval: ...
```

My reasoning is that we may want to add more properties in the future to configure the FetchService and having a consistent parent key across all plugins would be nice as an indicator to developers that a section of the config contains certain properties and uses the FetchService.

kjjuno replied:
added

kjjuno commented:
> Hey 👋
> 
> Apologies for the long turnaround on reviews for this, we've been pretty slammed after the holiday season, so we're just getting back to full capacity again!
> 
> I'm thinking that whilst we might get to having some form of `FetchService` in the backend, similar to the `fetchApiRef` in the frontend, we're not really there yet and we don't know what that's going to look like either. I don't want to be even more of a blocker for this work, so what I would suggest is that we narrow down the changes for what's required for this PR just to the `@backstage/plugin-bitbucket*` packages instead, and leave out the other changes for now.
> 
> It might need some restructuring of course, to maybe have this in some other config variable or to be set in some other way in the interim and out of the `integrations.config` for now, but I think that it's probably easier to go down that path, and then eventually deprecate that way once we have some more thoughts about what we're going to do with a global `FetchService`.
> 
> Does this make sense for now?

Totally makes sense. I'll restructure things to make that service only for the bitbucket plugins. I can also change the config as @joshjung suggested. I should have something up hopefully as early as tomorrow

joshjung commented:
> Totally makes sense. I'll restructure things to make that service only for the bitbucket plugins. I can also change the config as @joshjung suggested. I should have something up hopefully as early as tomorrow

Great! Once you have this done, what I can do is restructure the PR I have for the Gitlab plugin to match the same config format you use.

kjjuno commented:
This is ready for another look now I think

freben commented on docs/integrations/bitbucketCloud/locations.md:
```diff
@@ -20,6 +20,11 @@ integrations:
   bitbucketCloud:
     - username: ${BITBUCKET_CLOUD_USERNAME}
       appPassword: ${BITBUCKET_CLOUD_PASSWORD}
+      fetch: # optional
```
Might as well remove this nesting level (in the docs below it's not here)

joshjung replied:
This was something I specifically requested because we were introducing a new object that was going to be shared across multiple configurations for customizing fetch. Since I have to add it in another spot, would it not make sense to have the nesting so that there is a consistent indicator across configuration locations, since most likely this throttling will end up in a lot of spots in the future? What if we also want to add more configurations for fetching?

freben commented on docs/integrations/bitbucketServer/locations.md:
```diff
@@ -32,6 +37,11 @@ integrations:
       apiBaseUrl: https://bitbucket.mycompany.com/rest/api/1.0
       username: ${BITBUCKET_SERVER_USERNAME}
       password: ${BITBUCKET_SERVER_PASSWORD}
+      fetch: # optional
```
remove nesting

freben commented on packages/backend-defaults/src/entrypoints/urlReader/lib/BitbucketCloudUrlReader.ts:
```diff
@@ -26,13 +26,15 @@ import {
 import { NotFoundError, NotModifiedError } from '@backstage/errors';
 import {
   BitbucketCloudIntegration,
+  BitbucketFetchFunction,
+  BitbucketFetchService,
```
I really would want to get rid of these types. The core of it is, the integrations package should be left as pristine as is humanly possible. See comments below.

freben commented on packages/backend-defaults/src/entrypoints/urlReader/lib/BitbucketCloudUrlReader.ts:
```diff
@@ -26,13 +26,15 @@ import {
   getBitbucketCloudRequestOptions,
   ScmIntegrations,
 } from '@backstage/integration';
-import fetch, { Response } from 'node-fetch';
```
keep this import

freben commented on packages/backend-defaults/src/entrypoints/urlReader/lib/BitbucketCloudUrlReader.ts:
```diff
@@ -58,6 +60,8 @@ export class BitbucketCloudUrlReader implements UrlReaderService {
     });
   };
 
+  private readonly fetch: BitbucketFetchFunction;
```
```suggestion
  private readonly fetch: typeof fetch;
```

freben commented on packages/backend-defaults/src/entrypoints/urlReader/lib/BitbucketCloudUrlReader.ts:
```diff
@@ -69,6 +73,8 @@ export class BitbucketCloudUrlReader implements UrlReaderService {
       );
     }
+
+    this.fetch = BitbucketFetchService.get(this.integration.config);
```
Note the [integration-aws-node package](https://github.com/backstage/backstage/tree/526d5f1101a299cd0a9b0fe43c22355ef4a6bec3/packages/integration-aws-node). That integration got to a point where they realized that they couldn't stay within the confines of a polymorphic (web + node at the same time) package so their integration-specific code had to go in a package of its own. Crucially, _that package has a different role_ in its package.json, namely [node-library](https://github.com/backstage/backstage/blob/526d5f1101a299cd0a9b0fe43c22355ef4a6bec3/packages/integration-aws-node/package.json#L6) instead of common-library like the polymorpic ones. This is important because the build setup for those are entirely different and they are permitted to import node packages like node-fetch. And it let the regular integrations package stay slim.

The code that builds your special fetch function has to live in a node-library role package if it wants to be based on node-fetch. Unfortunately, bitbucket-cloud-common is not of that role. You have two options:

- First option: Rewrite your fetcher to be polymorphic, base it on only the cross-fetch library instead of node-fetch, and move its code into bitbucket-cloud-common. This would be kind of nice, but I expect that there might turn out to be some small differences between cross-fetch and node-fetch, making typescript upset when using that fetcher in your node packages.
- Second option: Create an integration-bitbucket-node package, make it of the node-library role, and move your implementation there. Then of course it won't be polymorphic.

The thing is, I think bitbucket-cloud-common should never have been a -common package in the first place. It is only used by backend packages, so making it polymorphic just made everything harder for no reason. I vote doing option 2, and (later down the line) deprecating bitbucket-cloud-common and moving all of its guts into the new integration-bitbucket-node package too. Don't do that last bit now though, we have enough complexity to handle anyway.

Adding one more thing. You want to be able to pass in the custom fetch function in the constructor (in the deps argument), instead of constructing it here. Then tests and alternative callers can pass in what they want. You may want to create the actual fetcher inside the factory above instead of the constructor.

freben commented on packages/backend-defaults/src/entrypoints/urlReader/lib/BitbucketServerUrlReader.ts:
```diff
@@ -57,10 +59,14 @@ export class BitbucketServerUrlReader implements UrlReaderService {
     });
   };
 
+  private fetch: BitbucketFetchFunction;
```
same comments here as in the other file

freben commented on packages/integration/api-report.md:
````diff
@@ -5,6 +5,10 @@
 ```ts
 import { Config } from '@backstage/config';
 import { ConsumedResponse } from '@backstage/errors';
+import { HumanDuration } from '@backstage/types';
````
this import is ok, the other ones are not

freben commented on packages/integration/api-report.md:
```diff
@@ -187,8 +191,15 @@ export type BitbucketCloudIntegrationConfig = {
   username?: string;
   appPassword?: string;
   token?: string;
+  fetch?: FetchConfig;
```
if you could change this to only be `throttling?: ThrottlingConfig` or so, then that's a small enough change to let through.

freben commented on packages/integration/api-report.md:
```diff
@@ -187,8 +191,15 @@ export type BitbucketCloudIntegrationConfig = {
+export type BitbucketFetchFunction = (
+  url: RequestInfo_2,
+  init?: RequestInit_2,
+) => Promise<Response_2>;
```
this has to be removed

freben commented on packages/integration/api-report.md:
```diff
@@ -301,6 +313,11 @@ export function defaultScmResolveUrl(options: {
+// @public
+export type FetchConfig = {
+  throttling?: ThrottlingConfig;
+};
```
remove

freben commented on packages/integration/package.json:
```diff
@@ -40,12 +40,16 @@
+    "@backstage/types": "workspace:^",
     "@octokit/auth-app": "^4.0.0",
     "@octokit/rest": "^19.0.3",
+    "@types/node-fetch": "^2.5.12",
```
remove

freben commented on packages/integration/package.json:
```diff
@@ -40,12 +40,16 @@
     "lodash": "^4.17.21",
-    "luxon": "^3.0.0"
+    "luxon": "^3.0.0",
+    "node-fetch": "^2.6.7",
+    "p-throttle": "^4.1.1"
```
remove

freben commented on packages/integration/src/bitbucket/BitbucketFetchService.ts:
```diff
@@ -0,0 +1,69 @@
+ * This service can be configured to throttle the number of requests that can be made.
+ * @internal
+ */
+export class BitbucketFetchService {
```
If you make a `@backstage/integration-bitbucket-node` package, it could have a

```ts
function createThrottledFetch(options: {
  inner: typeof fetch,
  throttleByHost: {
    [host in string]: ThrottleConfig;
  }
}): typeof fetch
```

or something like that, and make it like a "middleware-like" thing around the inner fetch, [like we do](https://github.com/backstage/backstage/blob/526d5f1101a299cd0a9b0fe43c22355ef4a6bec3/packages/core-app-api/src/apis/implementations/FetchApi/IdentityAuthInjectorFetchMiddleware.ts#L55) when we [create the fetchApi](https://github.com/backstage/backstage/blob/526d5f1101a299cd0a9b0fe43c22355ef4a6bec3/packages/core-app-api/src/apis/implementations/FetchApi/createFetchApi.ts#L36) for the frontend.

No need for caching fetch implementations in a map. Its internal state will be a `Map<string, typeof pThrottle>` instead.

freben commented on plugins/bitbucket-cloud-common/package.json:
```diff
@@ -42,11 +42,13 @@
     "@backstage/integration": "workspace:^",
-    "cross-fetch": "^4.0.0"
+    "cross-fetch": "^4.0.0",
+    "node-fetch": "^2.6.7"
```
remove

freben commented on plugins/bitbucket-cloud-common/package.json:
```diff
@@ -42,11 +42,13 @@
   "devDependencies": {
     "@backstage/cli": "workspace:^",
     "@openapitools/openapi-generator-cli": "^2.4.26",
+    "@types/node-fetch": "^2.5.12",
```
remove

freben commented on plugins/bitbucket-cloud-common/src/BitbucketCloudClient.ts:
```diff
@@ -32,9 +36,13 @@ export class BitbucketCloudClient {
     return new BitbucketCloudClient(config);
   }
 
+  private readonly fetch: BitbucketFetchFunction;
```
This one should also accept a `fetch?: typeof fetch` in its constructor. But we'll see if typescript gets cranky with the node-fetch vs cross-fetch types.

kjjuno replied:
typescript is complaining about this one. That was the motivation for creating the extra interface.

freben reviewed (commented):
Sorry for the delay and the large amount of comments. This turns out to be a very tricky thing, in part because of previous decisions that are none of your fault, but also because of just the tricky nature of the integrations package and polymorphism. That package is being revamped in some nearish future, making things a lot simpler. But here we are.

Let me summarize the comments:

- The -common packages cannot use node-fetch at all
- The integrations package needs to be kept almost perfectly pristine and must not contain any implementation code

The effect of this is:

- Need to make a node typed package, preferably `@backstage/integration-bitbucket-node` and have all implementation there
- Dial back the integrations package _even more_ to exclusively the throttler type change
- Change things to accept a `typeof fetch` making it entirely transparent to their implementations what the fetcher actually is doing.

kjjuno commented:
I am still working through this. I have been distracted by other priorities at work, but this is still very much on my mind and I intend to have an update posted next week sometime.

kjjuno commented:
Just now getting back to this after the weekend. Addressing the feedback is taking longer than expected. I'm hoping to have new changes up in the next couple of days

kjjuno commented:
I believe I have addressed all the feedback. I would be very grateful if you guys could take another look at this one

kjjuno commented:
Any chance I could get some eyes on this in the next few days?

kjjuno commented:
@freben @vinzscam @Rugvip @benjdlambert 

Is there any way I could get a review in the next couple of days?

freben commented:
Just to provide an update:

I've been through this one today and trying to move it forward. The main blocker I note, is the mixture of `common` and `node` type packages. I was trying to sort it out on my end to try to devise what the best way forward is, but ran out of time for the day at least.

The new package is named `@backstage/integration-bitbucket-node`, but it isn't a node role, but instead `common-library`. `common-library` is VERY special (since they are meant to slot perfectly into both browsers and node) and are absolutely forbidden from importing node concerns - importing `node-fetch` is a no-no.

At the same time, the pre-existing `@backstage/plugin-bitbucket-cloud-common` package now also gets `node-fetch` added, but also `@backstage/integration-bitbucket-node` (albeit as previously stated, just "misnamed" at the moment).

The thing is, the new facilities in `@backstage/integration-bitbucket-node` ARE imported both by `common-library` and `node-library` packages. So I tried rewriting it to be based on either the regular global `fetch` or the `cross-fetch` package, but it leads to interesting type errors because the two different environments who want to consume it have differently-behaving `Response` types etc.

I can dig a bit more when I get the time, but those things need to be sorted out. My feeling at the end of the day here, is maybe we could even just not introduce the `@backstage/integration-bitbucket-node` package at all, and instead introduce the `p-limit` thing in the few places it's needed instead (the url reader and the `@backstage/plugin-bitbucket-cloud-common` package somehow). Not sure.

freben commented:
Alright. As a basis for discussion, I've pushed a commit that detaches the new package from the actual fetch implementation. See how you like it, if it aligns with your own goals and ideas.

cal5barton commented:
> Alright. As a basis for discussion, I've pushed a commit that detaches the new package from the actual fetch implementation. See how you like it, if it aligns with your own goals and ideas.

The one question I have is, having the ThrottleService and ThrottleConfig defined in a bitbucket specific package requires all the other providers to do the same thing, is this the intention or is a better way to package up this common functionality for the other providers as well? #26063

freben commented:
Mm. One thing that could make it worth having a dedicated package would be if we could leverage actual server provided info. Ie if calls return rate limiting headers then we'd specifically obey those instead of having fixed limits. And those would be different for every provider. As it stands, the whole thing is a bit silly because there's so little reuse.

But yeah actually what would _really_ be needed is to have one single shared instance of this limiter per host. Creating separate ones in several places (like the url reader and entity provider) isn't really acceptable if we're going to be serious about really obeying the limits. That would normally entail making a [service ref](https://backstage.io/docs/backend-system/architecture/services) for it on the backend side (in a -node package, not common!) and then injecting that wherever needed.

And then the next step would be to have a shared backing store for all of those, like redis, so that your entire backstage installation as a whole can obey global limits.

cal5barton commented:
@freben @vinzscam @benjdlambert @Rugvip,

What is the direction forward on this issue? It seems there is no clear path, and @kjjuno and I have been trying to advance this since May.

Due to increasing pressure to have this visibility in our organization, we need to make progress in our instance. If you can provide additional clarity on the direction (as several other plugins require this same functionality), we would be happy to contribute. Alternatively, if you prefer to take over, that would be acceptable as well.

Thanks!

Rugvip commented:
@cal5barton would you be able to join the [framework SIG](https://github.com/backstage/community/tree/main/sigs/sig-framework) meeting next week? I think a bit of synchronous discussion can help move this forward more quickly as we still find ourselves stuck with a lot of open questions

cal5barton commented:
> @cal5barton would you be able to join the [framework SIG](https://github.com/backstage/community/tree/main/sigs/sig-framework) meeting next week? I think a bit of synchronous discussion can help move this forward more quickly as we still find ourselves stuck with a lot of open questions

Yes @kjjuno and I will try to attend that meeting.

kjjuno commented:
As discussed in the sig, this PR is getting deprecated in favor of this more targeted PR.

https://github.com/backstage/backstage/pull/27850

hughbidone commented:
Thanks so shall we close this and focus on https://github.com/backstage/backstage/pull/27850
